samza-azure/src/main/java/org/apache/samza/system/eventhub/admin/EventHubSystemAdmin.java

```diff
@@ -0,0 +1,164 @@
+import org.apache.samza.system.eventhub.EventHubClientWrapperFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class EventHubSystemAdmin implements SystemAdmin {
+  private static final Logger LOG = LoggerFactory.getLogger(EventHubSystemAdmin.class);
+  private final EventHubClientWrapperFactory eventHubClientWrapperFactory;
+  private String systemName;
+  private EventHubConfig eventHubConfig;
+  private Map<String, EventHubClientWrapper> eventHubClients = new HashMap<>();
+  private Map<String, String[]> streamPartitions = new HashMap<>();
+
+  public EventHubSystemAdmin(String systemName, EventHubConfig eventHubConfig,
+                             EventHubClientWrapperFactory eventHubClientWrapperFactory) {
+    this.systemName = systemName;
+    this.eventHubConfig = eventHubConfig;
+    this.eventHubClientWrapperFactory = eventHubClientWrapperFactory;
+  }
+
+  private static String getNextOffset(String currentOffset) {
+    // EventHub will return the first message AFTER the offset
+    // that was specified in the fetch request.
+    return currentOffset.equals(EventHubSystemConsumer.END_OF_STREAM) ? currentOffset :
+            String.valueOf(Long.parseLong(currentOffset) + 1);
+  }
+
+  @Override
+  public Map<SystemStreamPartition, String> getOffsetsAfter(Map<SystemStreamPartition, String> offsets) {
+    Map<SystemStreamPartition, String> results = new HashMap<>();
+    offsets.forEach((partition, offset) -> results.put(partition, getNextOffset(offset)));
+    return results;
+  }
+
+  @Override
+  public Map<String, SystemStreamMetadata> getSystemStreamMetadata(Set<String> streamNames) {
+    Map<String, SystemStreamMetadata> requestedMetadata = new HashMap<>();
+    Map<String, CompletableFuture<EventHubRuntimeInformation>> ehRuntimeInfos = new HashMap<>();
+    streamNames.forEach((streamName) -> {
+        if (!streamPartitions.containsKey(streamName)) {
+          EventHubClientWrapper eventHubClientWrapper = getStreamEventHubClient(streamName);
+          ehRuntimeInfos.put(streamName,
+                  eventHubClientWrapper.getEventHubClient().getRuntimeInformation());
+        }
+      });
+    ehRuntimeInfos.forEach((streamName, ehRuntimeInfo) -> {
+        if (!streamPartitions.containsKey(streamName)) {
+          try {
+            EventHubRuntimeInformation ehInfo = ehRuntimeInfo.get(eventHubConfig.getRuntimeInfoWaitTimeMS(),
+                    TimeUnit.MILLISECONDS);
+            streamPartitions.put(streamName, ehInfo.getPartitionIds());
+          } catch (InterruptedException | ExecutionException e) {
+            String msg = String.format("Error while fetching EventHubRuntimeInfo for System:%s, Stream:%s",
```

Comment:
Cant we use the same path for all exceptions?  We appear to be handling exceptions the same way in either of these `catch` code paths.